ansible_base/activitystream/models/entry.py

```diff
@@ -0,0 +1,83 @@
+from functools import partial
+
+from django.contrib.contenttypes.fields import GenericForeignKey
+from django.contrib.contenttypes.models import ContentType
+from django.db import models
+from django.db.models.signals import m2m_changed, post_save, pre_delete, pre_save
+from django.utils.translation import gettext_lazy as _
+
+from ansible_base.activitystream.signals import activitystream_create, activitystream_delete, activitystream_update, activitystream_m2m_changed
+from ansible_base.lib.abstract_models import CommonModel
+
+
+class Entry(CommonModel):
+    """
+    An activity stream entry.
+
+    This is keyed on a generic object_id and content_type, which allows for
+    a wide variety of objects to be used in the activity stream.
+    """
+
+    allow_anonymous_user_save = True
+
+    class Meta:
+        verbose_name_plural = _('Entries')
+
+    OPERATION_CHOICES = [
+        ('create', _('Entity created')),
+        ('update', _("Entity updated")),
+        ('delete', _("Entity deleted")),
+        ('associate', _("Entity was associated with another entity")),
+        ('disassociate', _("Entity was disassociated with another entity")),
+    ]
+
+    content_type = models.ForeignKey(ContentType, on_delete=models.DO_NOTHING)
+    object_id = models.TextField(null=True, blank=True)
+    content_object = GenericForeignKey('content_type', 'object_id')
+    operation = models.CharField(max_length=12, choices=OPERATION_CHOICES)
+    changes = models.JSONField(null=True, blank=True)
+
+    # This is used for m2m (dis)associations
+    related_content_type = models.ForeignKey(ContentType, on_delete=models.DO_NOTHING, null=True, blank=True, related_name='related_content_type')
+    related_object_id = models.TextField(null=True, blank=True)
+    related_content_object = GenericForeignKey('related_content_type', 'related_object_id')
+
+    # TODO: AWX stores denormalized actor data to account for deleted users, we should do the same
+
+    def __str__(self):
+        return f'[{self.created_on}] {self.get_operation_display()} by {self.created_by}: {self.content_type} {self.object_id}'
+
+
+class AuditableModel(models.Model):
+    """
+    A mixin class that provides integration to the activity stream from any
+    model. A model should simply inherit from this class to have its
+    create/update/delete events sent to the activity stream.
+    """
+
+    class Meta:
+        abstract = True
+
+    activity_stream_excluded_field_names = []
+
+    @classmethod
+    def connect_signals(cls):
```

Comment:
I'd rather not have this as a class method, since there's no need for the nonsense call of `Organization.connect_signals`. That doesn't even clarify _what_ signals. If someone does `dir(organization)` this adds cruft to what they see. You could leave this as a plain function lying around wherever in this activity stream app.